File: Challenges/450_DSA_Cracker/010_Min_Jumps_Across_Array.java
//{ Driver Code Starts
import java.lang.*;
import java.io.*;
import java.util.*;
class GFG
 {
	public static void main (String[] args) throws IOException
	 {
	 
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 
        int t = Integer.parseInt(br.readLine()); 

        while(t-- > 0){
            int size = Integer.parseInt(br.readLine());
            String[] arrStr = ((String)br.readLine()).split("\\s+");
            int[] arr= new int[size];
            for(int i = 0;i<size;i++){
                arr[i] = Integer.parseInt(arrStr[i]);
            }
            System.out.println(new Solution().minJumps(arr));
        }
	 }
	 
}

// } Driver Code Ends


class Solution{
    static int minJumps(int[] arr){
        
        int l = arr.length;
        int reach = arr[0];
        int ind = 0;
        int steps = 1;
        
        while (reach < l-1) {
            int mx = 0;
            int mxind = -1;
            for(int i = ind+1; i < ind+1+arr[ind]; i++) {
                if(i+1+arr[i] > mx) {
                    mx = i+arr[i];
                    mxind = i;
                }
            }
            
            if(mx <= reach) {
                return -1;
            }
            
            reach = mx;
            ind = mxind;
            steps++;
        }
        
        return steps;
    }
}

File: Challenges/450_DSA_Cracker/Array/010_Min_Jumps_Across_Array.py
#User function Template for python3
class Solution:
    def minJumps(self, arr, n):
        steps = 1
        step_index = 0
        reach = step_index+arr[step_index]
        
        while (reach+1) <= n :
            mx = 0
            mxind = -1
            for i in range(step_index+1, reach+1) :
                if i+arr[i] > mx : 
                    mx = i+arr[i]
                    mxind = i
            if mx <= reach :
                return -1
            if reach+1 >= n : 
                break
            else : 
                # print("Step {} from {}. Reach = {}".format(arr[mxind], mxind, reach))
                steps += 1
                step_index = mxind
                reach = step_index + arr[step_index]
        
        return steps


#{ 
 # Driver Code Starts
#Initial Template for Python 3
if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		n = int(input())
		Arr = [int(x) for x in input().split()]
		ob = Solution()
		ans = ob.minJumps(Arr,n)
		print(ans)
# } Driver Code Ends
